Viewer Snapshot to Catalogue

ericmehl opened the pull request:
This does pretty much what the title suggests. In addition to sending the pixel data, a display window indicating the camera's resolution gate will be added if the snapshot is taken when viewing through a camera. I'm currently including the frame and the scene plug from which the snapshot was taken in metadata.

Identifying the catalogues in a script is one area that might deserve some more attention. I'm using `GafferImage.Catalogue.RecursiveRange()` to find all `Catalogue` nodes from the script root. Some recent production scripts have around 600k - 1m nodes. Remote machines didn't seem to have any noticeable lag in getting all of the catalogues using  `list( catalogues )` in the Python terminal.

I do notice a pretty good lag on my workstation with an artificially generated network of a similar number of nodes - 3-4 seconds on a release build.

Not sure if that's a matter of the topology of the node graphs or simply CPU speed / cores (I have 48, not sure what the remote machines have).

I put the `Catalogue` search into the "Send to Catalogue" submenu so it can be deferred until the user expands that menu, so that provides a bit of a guard against unwanted lags.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon commented on include/GafferSceneUI/Private/OutputBuffer.h:
```diff
@@ -81,6 +81,9 @@ class OutputBuffer
 		using BufferChangedSignal = Gaffer::Signals::Signal<void()>;
 		BufferChangedSignal &bufferChangedSignal();
 
+		const Imath::Box2i &dataWindow() const;
+		const std::vector<float> &rgbaBuffer() const;
```
Exposing these directly is dangerous, because the buffers can be modified from a background thread at any time - see `OutputBuffers::m_bufferReallocationMutex`. Unless we have other uses for exposing the buffers directly, and a mechanism for protecting that, I think I'd suggest we replace them with a `BufferAlgo::snapshotToFile()` method which we forward to directly from `SceneGadget::snapshotToFile()`.

ericmehl replied:
I removed those in https://github.com/GafferHQ/gaffer/pull/5533/commits/5dead3ccef2b62c176f6f43e542c7f16794a6df6 and shifted the implementation to `OutputBuffer::snapshotToFile()` in that commit and https://github.com/GafferHQ/gaffer/pull/5533/commits/b6c2f5367845b1fe6f4d802a0ea795e2c95c9fa6. Should it actually be in `BufferAlgo::snapshotToFile()`? `BufferAlgo` seems more about image window tools than handling the buffer lock and such, so I put it in `OutputBuffer` for now.

johnhaddon replied:
Sorry, yep, I meant `OutputBuffer` rather than `BufferAlgo` - thanks for translating my gibberish into something sensible.

johnhaddon commented on Changes.md:
```diff
@@ -14,6 +14,7 @@ API
 ---
 
 - Display : Added `driverClosed()` method.
+- SceneGadget : Add `snapshotToFile()` method.
```
Add -> Added

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5533/commits/6f11f56809664346e00e959e968ce0992adc28ff

johnhaddon commented on include/GafferSceneUI/SceneGadget.h:
```diff
@@ -191,6 +193,16 @@ class GAFFERSCENEUI_API SceneGadget : public GafferUI::Gadget
 		/// Implemented to return the name of the object under the mouse.
 		std::string getToolTip( const IECore::LineSegment3f &line ) const override;
 
+		/// Saves a snapshot of `m_outputBuffer` to the given file. All formats supported by
```
`m_outputBuffer` is a private member and a private class, so isn't really appropriate for API docs. Can we talk more generically about the rendered image, and explain that it's not available for the OpenGL renderer?

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5533/commits/b6c2f5367845b1fe6f4d802a0ea795e2c95c9fa6

johnhaddon commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -780,6 +784,51 @@ Imath::Box3f SceneGadget::bound() const
+
+	const std::vector<float> &rgbaBuffer = m_outputBuffer->rgbaBuffer();
+
+	const IECoreImage::OpenImageIOAlgo::DataView dataView( metadata );
```
This `dataView` is unused as far as I can see, and shadowed by the one inside the loop.

ericmehl replied:
Fixed amongst https://github.com/GafferHQ/gaffer/pull/5533/commits/b6c2f5367845b1fe6f4d802a0ea795e2c95c9fa6 and https://github.com/GafferHQ/gaffer/pull/5533/commits/5dead3ccef2b62c176f6f43e542c7f16794a6df6

johnhaddon commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -780,6 +784,51 @@ Imath::Box3f SceneGadget::bound() const
+		}
+	}
+
+	std::unique_ptr<OIIO::ImageOutput> output = OIIO::ImageOutput::create( fileName.c_str() );
```
I wonder if we should be writing half-float images when the format is EXR? That's what we write when an ImageNode is dragged to the Catalogue. Any opinion @murraystevenson?

ericmehl replied:
That sounds good to me, I suppose full float data for the viewport is a little excessive. I'll defer to @murraystevenson for a final go or no go.

johnhaddon commented on src/GafferSceneUIModule/SceneGadgetBinding.cpp:
```diff
@@ -214,6 +214,7 @@ void GafferSceneUIModule::bindSceneGadget()
 		.def( "getSelection", &SceneGadget::getSelection, return_value_policy<copy_const_reference>() )
 		.def( "selectionBound", &selectionBound )
 		.def( "bound", &bound, ( arg( "selected" ), arg( "omitted" ) = object() ) )
+		.def( "snapshotToFile", &SceneGadget::snapshotToFile )
```
Might be worth releasing the GIL for this, since it's a relatively heavy operation?

johnhaddon commented on Changes.md:
```diff
@@ -1,6 +1,11 @@
+Features
+--------
+
+- Viewer : Added the ability to send a snapshot of the current viewport to a catalogue.
```
Might be worth mentioning that this is available from the right-click menu?

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5533/commits/d2245108847a725a76943d32f84d38b3cff59418

johnhaddon commented on python/GafferSceneUI/SceneViewUI.py:
```diff
@@ -1073,13 +1076,105 @@ def __appendClippingPlaneMenuItems( menuDefinition, prefix, view, parentWidget )
+	timeStamp = str( datetime.datetime.now() )
+
+	fileName = pathlib.Path(
+		view["in"].node().getContext().substitute( catalogue["directory"].getValue() )
```
Took me a while to figure out what was happening here, but `view["in"].node()` is just `view`, right? I think I'd be tempted to use `scriptRoot.context()` instead - there's a fair chance that these individual context accessors on UI components will disappear over the long term.

ericmehl replied:
Good to know about trending away from UI component contexts, and a big "duh" to myself regarding `view["in"].node()` being `view`. Fixed in https://github.com/GafferHQ/gaffer/pull/5533/commits/792babe29290021ed8936aafab31039f06386bfe

johnhaddon commented on python/GafferSceneUI/SceneViewUI.py:
```diff
@@ -1073,13 +1076,105 @@ def __appendClippingPlaneMenuItems( menuDefinition, prefix, view, parentWidget )
+
+	fileName = pathlib.Path(
+		view["in"].node().getContext().substitute( catalogue["directory"].getValue() )
+	) / ( timeStamp.replace( ":", "-" ).replace(" ", "_" ) + ".exr" )
```
Should we use a `viewerSnapshot` prefix or something like that in the filename, to give a bit more of a clue about where the file came from?

ericmehl replied:
Added in https://github.com/GafferHQ/gaffer/pull/5533/commits/792babe29290021ed8936aafab31039f06386bfe

johnhaddon commented on python/GafferSceneUI/SceneViewUI.py:
```diff
@@ -1073,13 +1076,105 @@ def __appendClippingPlaneMenuItems( menuDefinition, prefix, view, parentWidget )
+		)
+	)
+
+	catalogue["images"].source().addChild( image )
```
Having played with this a little bit, I found myself with the same expectation as Murray - that this would also select the image in the catalogue.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5533/commits/2798d8be7d061bf93d2b5eff1088444988a19d9e

johnhaddon commented on python/GafferSceneUI/SceneViewUI.py:
```diff
@@ -1073,13 +1076,105 @@ def __appendClippingPlaneMenuItems( menuDefinition, prefix, view, parentWidget )
+		for c in catalogueList :
+
+			cName = c["name"].getValue()
+			nName = c.relativeName( scriptNode )
```
When a Catalogue has been put in a Box, Reference or custom node and then promoted (which is very common), this is giving the name of the internal Catalogue node which might well be private. I think we want to use the name of the promoted-to node in this case, which can be done pretty easily with `catalogue["imageIndex"].source().node()`.

johnhaddon commented on python/GafferSceneUI/SceneViewUI.py:
```diff
@@ -1073,13 +1076,105 @@ def __appendClippingPlaneMenuItems( menuDefinition, prefix, view, parentWidget )
+			menuDefinition.append(
+				"/" + ( nName + ( " ({})".format( cName ) if cName else "" ) ),
+				{
+					"active" : snapshotActive,
```
`and not Gaffer.MetadataAlgo.readOnly( c )`.

ericmehl replied:
I think it should be even more specific to `c["imageIndex"].source()` and `c["images"].source()`? I'm doing that in https://github.com/GafferHQ/gaffer/pull/5533/commits/3d73663775ffcc85463d4b23c8b8f35229586c57

I think if we just look at `c` then any time the catalog is in a Reference it won't be editable, correct?

johnhaddon replied:
Yep, spot on. Thanks once again for translating my gibberish into something mergeable.

johnhaddon reviewed (changes requested):
Thanks Eric, this is a nice feature! I've requested some changes inline, but hopefully nothing too drastic...
Cheers...
John

ericmehl commented on python/GafferSceneUI/SceneViewUI.py:
```diff
@@ -1073,13 +1076,105 @@ def __appendClippingPlaneMenuItems( menuDefinition, prefix, view, parentWidget )
+		)
+	)
+
+	catalogue["images"].source().addChild( image )
```
Somewhere here I need to set the current image in the catalog to be the one just added

ericmehl commented:
I've addressed all the comments save the float-to-half conversion. I also added a bit more information to the tooltip in the menu for catalogs that are disabled because their plug is not editable in https://github.com/GafferHQ/gaffer/pull/5533/commits/ae00e627b1c0b1ba70538aebabc0c42595d9d616. Hopefully that can save users a little bit of hunting for the source of the menu item being inactive.

johnhaddon commented:
Thanks Eric, LGTM. Could you get it all squashed ready to merge and tack a final half-floatifying commit on the end in anticipation please? Bear in mind that we've made a release (or two?) since the PR was opened, to the Changes.md bits all need shuffling up into the latest section.

ericmehl commented:
Changed the output type to half-float ✔️ , squashed ✔️  and moved the `changes.md` entries up to the top ✔️ . Should be ready to go.

johnhaddon commented:
Thanks Eric - I've rebased just one more time to fix the inevitable Changes.md conflict. Merging...
